### ISSUE TITLE
distributed training with c10d does not work with layerdrop in pytorch > 1.4

### ISSUE THREAD

**Issue by alexeib** [pytorch/pytorch#42793]
training any model that uses layerdrop (probabilistically skip transformer layers during training) works poorly with c10d distributed training backend (using fairseq). in pytorch 1.4 training would freeze, but if you specified --find-unused-parameters, then it would work. in pytorch 1.5 and above, the training always crashes with an error about inconsistent gradients.

to repro try any transformer based model in fairseq with the c10d backend and layerdrop > 0

for example, for wav2vec 2.0 (from the readme):

python train.py --distributed-world-size 64 --distributed-port $PORT /manifest/path \
--save-dir /model/path fp16 --num-workers 6 --task audio_pretraining --criterion wav2vec --arch wav2vec2 \
--log-keys '["prob_perplexity","code_perplexity","temp"]' --quantize-targets --extractor-mode default \
--conv-feature-layers '[(512, 10, 5)] + [(512, 3, 2)] * 4 + [(512,2,2)] * 2' --final-dim 256 --latent-vars 320 \
--latent-groups 2 --latent-temp '(2,0.5,0.999995)' --infonce --optimizer adam \
--adam-betas '(0.9,0.98)' --adam-eps 1e-06 --lr-scheduler polynomial_decay --total-num-update 400000 \
--lr 0.0005 --warmup-updates 32000 --mask-length 10 --mask-prob 0.65 --mask-selection static --mask-other 0 \
--encoder-layerdrop 0.05 --dropout-input 0.1 --dropout-features 0.1 --feature-grad-mult 0.1 \
--loss-weights '[0.1, 10]' --conv-pos 128 --conv-pos-groups 16 --num-negatives 100 --cross-sample-negatives 0 \
--max-sample-size 250000 --min-sample-size 32000 --dropout 0.1 --attention-dropout 0.1 --weight-decay 0.01 \
--max-tokens 1400000 --max-update 400000 --skip-invalid-size-inputs-valid-test

try with: --find-unused-parameters with pytorch 1.4, or --backend no_c10d (pytorch 1.5+) to get this to actually work

**Comment by mrshenli** [comment]
This might be caused by #28883 (cc @baobablyh). Since v1.5, DDP backward no longer touches a local `param.grad` if it is not defined by the local autograd engine, which could happen when you skip layers. And we need this feature to avoid messing up with some optimizer states.

@ngoyal2707 is this a known issue to you as well? Does fairseq have a fix for this?

**Comment by mrshenli** [comment]
@alexeib could you please also post this issue to fairseq repo? If it is indeed the None grad problem, the fix might need to be added to fairseq.

**Comment by alexeib** [comment]
@mrshenli i am one of fairseq maintainers so i can help fix it there if needed. what is the suggested fix, initialize all grads to empty tensors?

also have you considered adding some meaningful error messages when c10d fails in this way, including with --find-unused-parameters. currently it is not really possible what parameters are unused or why (i hacked fairseq to print out params with empty grads when using fairseq's no_c10d backend) or how to fix it

**Comment by mrshenli** [comment]
@alexeib the comments below are based on the assumption that the error is indeed caused by `None` grads, not 100% sure if this is the case. :) BTW, how does the grad consistency check in fairseq deal with None grads? And for the failed cases, are the grads indeed numerically different?

> initialize all grads to zero tensors?

This would work for some use cases, but not all. There might be some model using `param.grad==None` to decide whether it should skip weight decay in one iteration. See more discussion in #28294. I wonder if we should just skip None grads.

> also have you considered adding some meaningful error messages when c10d fails in this way

I might miss sth, but IIUC, the grad consistency check is not a feature in c10d DDP. In this case, how can we detect this error in c10d?


> including with --find-unused-parameters. currently it is not really possible what parameters are unused or why (i hacked fairseq to print out params with empty grads when using fairseq's no_c10d backend) or how to fix it

good point! we can detect and print out unused parameters in the forward pass, any suggestion on how this should be presented? Maybe `module_name.sub_module_name.param_name`? 

The hack you mentioned is a commonly used way to find unused parameters when debugging and you can run it on the local model without enabling distributed training as below. I would like to learn more about the use case. Is there any reason that it is hard to add the following two lines in the application side and have to hack into fairseq?

```python
local_model(inputs).sum().bacwkard()
print(lambda p: p.grad == None, local_model.parameters())
```

**Comment by alexeib** [comment]
thanks for the reply @mrshenli!

fairseq consistency checking works by computing norm of all gradients for all workers and then comparing it across workers (within some threshold). the grads are collected like so:
```
grads = [p.grad.detach() for p in filter(lambda p: p.grad is not None, params)]
```

so parameters with None grads are skipped. however, with fairseq's no_c10d mode, grads are stored in one big flat tensor where None grads are replaced with 0s so in this case everything works fine.

>This would work for some use cases, but not all. There might be some model using param.grad==None to decide whether it should skip weight decay in one iteration. See more discussion in #28294. I wonder if we should just skip None grads.

maybe these kinds of use cases are better handled by modifying the optimizer instead of relying on side-effects? or at least print/raise some sort of warning if this is seen?

anyway, currently the behavior with things like layerdrop is that either a) training freezes (torch 1.4 without --find-unused-parameters) or b) inconsistent gradients, and fairseq suggests using no_c10d

> good point! we can detect and print out unused parameters in the forward pass, any suggestion on how this should be presented? Maybe module_name.sub_module_name.param_name?

yes that would be great! (maybe only print those if --find-unused-parameters is omitted)

**Comment by alexeib** [comment]
zeroing out grads when they were missing before/after backward() did not work :(
any other ideas?

**Comment by mrshenli** [comment]
Hey @alexeib can you confirm that grads are consistent when not using layer-drop? 

> any other ideas?

hmm, not on top of my mind, but it looks like a real bug that we need to fix if it is not the None grad issue. @rohan-varma have you seen similar errors when working on the model consistency tests?

> yes that would be great! (maybe only print those if --find-unused-parameters is omitted)

we might need to enable some debugging mode/ctx for this, as doing this detection by default would incur extra overhead.

**Comment by alexeib** [comment]
thanks @mrshenli !

yes, if i disable layer drop everything works great. if i use no_c10d backend everything also works fine. if i use c10d without --find-unused-parameters then the training freezes. if i use --find-unused-parameters then it fails with inconsistent gradients error. happy to test with a debug build or provide a repro command

**Comment by mrshenli** [comment]
adding a high-pri tag to make sure we get this fixed before the next release.

**Comment by rohan-varma** [comment]
> hmm, not on top of my mind, but it looks like a real bug that we need to fix if it is not the None grad issue. @rohan-varma have you seen similar errors when working on the model consistency tests?
> 

I haven't seen this in model consistency tests. Does this reproduce on any model that probabilistically skips layers (i.e. could we write a simple unittest to reproduce this)?

**Comment by alexeib** [comment]
i believe so. here's where it happens in the model that is breaking for me, but it should be analogous for any other simpler model:

https://github.com/pytorch/fairseq/blob/522c76ba1646cd5ec2cd4be29392f53d40aec50a/fairseq/models/wav2vec/wav2vec2.py#L827
